Gemfile

```diff
@@ -252,7 +252,13 @@ end
 def override_gem(name, *args)
   if dependencies.any?
     raise "Trying to override unknown gem #{name}" unless (dependency = dependencies.find { |d| d.name == name })
-    dependencies.delete(dependency)
+
+    removed_dependency = dependencies.delete(dependency)
+    if removed_dependency.source.kind_of?(Bundler::Source::Git)
+      @sources.send(:source_list_for, removed_dependency.source).delete_if do |other_source|
+        @sources.send(:equal_source?, removed_dependency.source, other_source)
```

Comment:
Wow, nice find.  Is there a public api for this?  Note, we're modifying the array in place with `delete_if`, right? ... if they freeze it in the future, I believe this would raise an exception.

Reply:
I figured out the code for this a while ago for `bisect_gemfile_parser.rb` in NickLaMuro/miq_tools so I would have to look again tomorrow.  My guess is not, since `bundler` tends to not be good about public API's for this kind of thing, though, there private APIs tend to be pretty stable for the most part (at least in my experience).

Reply:
> Is there a public api for this?

Turns out that `.equal_source?` is not available on `bundler v1.15.4`, and was added in `1.16.x`, and we use `1.15.4` for builds, so this is a hard requirement that we work around this.  Turns out `.equal_source?` is pretty much `source == other_source`, so I will be switching to that as a result.

> Note, we're modifying the array in place with delete_if, right? ... if they freeze it in the future, I believe this would raise an exception.

From what I am looking at, thee doesn't seem to be a public API, but I will check back after doing some more digging.  I am at least going to change the above to be:

```ruby
removed_dependency.source == other_source
```

So it is at least a bit more portable across `bundler` versions.

Reply:
> ... if they freeze it in the future, I believe this would raise an exception.

So I have been looking into this for a good part of today, and I don't think there will be an issue with how `Bundler` treats it's `Bundler::SourceList` as pretty fluid since it's inception:

https://github.com/bundler/bundler/commit/edccf04e
https://github.com/bundler/bundler/blob/edccf04e/lib/bundler/source_list.rb#L8-L9

However, there is a big issue I have noticed with that, though.  `Bundler::SourceList` instances will call a `.uniq!` on any source that is added to a particular list via `add_source_to_list`:

https://github.com/bundler/bundler/blob/edccf04e/lib/bundler/source_list.rb#L57-L60

Since this is doing a destructive operation to the list, existing dependencies might loose their `@source` reference to their source in the main list that is attached to them on instantiation:

https://github.com/bundler/bundler/blob/edccf04e/lib/bundler/dependency.rb#L43

And because this `Bundler::SourceList` implementation doesn't seem to be particularly tied to the dependencies it is hooked into, it doesn't do a check before removal (via `.uniq!`, or what I am doing here).

* * *

This is a long winded explanation to say that for our purposes, this #shouldBeFine™, but this is using pretty loose APIs that aren't defined the best, and I don't know that there is a better option available.

That said, I might try and break this out into a [`bundler plugin`](https://bundler.io/v1.17/guides/bundler_plugins.html) (available since `1.14.x`, which might be a better way of doing this.  We would still have to "DSL injection", and there is no clean way to do that with their plugin API, but we could have this functionality much better tested and edge cases like this tested against in multiple versions of `bundler`, compared to now where we are finding out if this works for certain versions using the MIQ build process.

Reply:
> and we use 1.15.4 for builds

I don't see why we can't bump the minimum unless @simaishi has a reason.

Reply:
> ...unless @simaishi has a reason.

@Fryguy She gave me one (though, I didn't request a link), and said that you and @jrafanie had a discussion about it 😉 

That said, the change for `equal_source?` was kind of an esthetic tacked on to almost completely unrelated change, so I don't think that using it is a big deal.  `.source_list_for` is the other API method I am using, and that seems like it has been a bit more stable since it hasn't really changed in years.

Reply:
> > ...unless @simaishi has a reason.
> 
> @Fryguy She gave me one (though, I didn't request a link), and said that you and @jrafanie had a discussion about it 😉

When I initially tried bumping to 1.16.x (this is a while ago....), after updating bundler, `bundle check` was no longer able to find some gems.  Here are the steps taken:

1. On a downstream appliance, run `bundle check` - everything is good
1. Install a newer version of bundler
1. Run `bundle check` again - this will list all system gems as missing 

When I say "system gems", those are the gems installed to system path (which is where gems get installed for any compiled gems that we build RPM for).

I looked for configuration changes that can be made to make it work, but didn't find any. Our options were either to rebuild all gem RPMs to install to where other gems are, or stay with 1.15.x and decision at that time was the latter.